### PR TITLE
feat: light node

### PR DESCRIPTION
**Pull request by mfrankovi**
Implements light node functionality by deleting old data from period_data column which takes most of the disk space in the database.

light-node-history parameter in the config file defines how much period_data history is kept in the database. The rest is deleted.

As part of this a new feature of limiting inclusion of old/expired dag blocks into pbft is implemented.

### CODE REVIEW

File: libraries/cli/include/cli/default_config.hpp [hunk 1]
```diff
@@ -16,6 +16,8 @@ const char *default_json = R"foo({
   "network_sync_level_size": 10,
   "network_packets_processing_threads": 14,
   "network_peer_blacklist_timeout" : 600,
+  "network_peer_light_node_timeout" : 600,
+  "light-node-history" : 0,
```
Comment (MatusKysel) [inline libraries/cli/include/cli/default_config.hpp:20]: naming

File: libraries/core_libs/consensus/src/dag/dag.cpp [hunk 2]
```diff
@@ -418,6 +418,14 @@ void DagManager::addDagBlock(DagBlock const &blk, SharedTransactions &&trxs, boo
           LOG(log_dg_) << "Block already in DB: " << blk.getHash();
           return;
         }
+        const auto proposal_period = dag_blk_mgr_->getProposalPeriod(blk.getLevel());
+        assert(proposal_period.second);
+        if (proposal_period.first + FullNodeConfig::kDagExpiryPeriodLimit < period_) {
```
Comment (MatusKysel) [inline libraries/core_libs/consensus/src/dag/dag.cpp:423]: same check could be part of `verifyBlock` as we do not need to verify block if it is too old

File: libraries/core_libs/consensus/src/dag/dag.cpp [hunk 3]
```diff
@@ -576,25 +588,89 @@ uint DagManager::setDagBlockOrder(blk_hash_t const &new_anchor, uint64_t period,
...
-        auto dag_block = dag_blk_mgr_->getDagBlock(blk);
+    for (auto &blk_hash : v.second) {
+      if (dag_order_set.count(blk_hash) == 0) {
+        auto dag_block = dag_blk_mgr_->getDagBlock(blk_hash);
         auto pivot_hash = dag_block->getPivot();
-        addToDag(blk, pivot_hash, dag_block->getTips(), dag_block->getLevel(), false);
+
+        if (validateBlockNotExpired(dag_block, expired_dag_blocks_to_remove)) {
+          addToDag(blk_hash, pivot_hash, dag_block->getTips(), dag_block->getLevel(), false);
+        }
```
Comment (MatusKysel) [inline libraries/core_libs/consensus/src/dag/dag.cpp:601]: I guess this should be only called when we are light node

File: libraries/core_libs/network/rpc/Test.cpp [hunk 4]
```diff
@@ -47,11 +47,17 @@ Json::Value Test::get_dag_block(const Json::Value &param1) {
...
     if (auto node = full_node_.lock()) {
       blk_hash_t hash = blk_hash_t(param1["hash"].asString());
-      auto blk = node->getDagBlockManager()->getDagBlock(hash);
-      if (!blk) {
-        res = "Block not available \n";
+      const auto &dag_blk_mgr = node->getDagBlockManager();
+      auto known_block = dag_blk_mgr->isDagBlockKnown(hash);
+      if (known_block) {
+        auto blk = dag_blk_mgr->getDagBlock(hash);
+        if (!blk) {
```
Comment (MatusKysel) [inline libraries/core_libs/network/rpc/Test.cpp:54]: I do not understand this change, why not call `getDagBlock` as `isDagBlockKnown` calls inside `getDagBlock` so you basically calling it 2x

File: libraries/core_libs/network/src/tarcap/packets_handlers/status_packet_handler.cpp [hunk 5]
```diff
@@ -51,6 +52,17 @@ void StatusPacketHandler::process(const PacketData& packet_data, const std::shar
     auto const node_major_version = (*it++).toInt<unsigned>();
     auto const node_minor_version = (*it++).toInt<unsigned>();
     auto const node_patch_version = (*it++).toInt<unsigned>();
+    uint64_t node_history = (*it++).toInt<uint64_t>();
```
Comment (MatusKysel) [inline libraries/core_libs/network/src/tarcap/packets_handlers/status_packet_handler.cpp:55]: const

File: libraries/core_libs/network/src/tarcap/packets_handlers/status_packet_handler.cpp [hunk 6]
```diff
@@ -51,6 +52,17 @@ void StatusPacketHandler::process(const PacketData& packet_data, const std::shar
...
+
+    // If this is a light node and it cannot serve our sync request disconnect from it
+    if (node_history > 0) {
+      if (pbft_synced_period + node_history < peer_pbft_chain_size) {
+        LOG(log_nf_) << "Light node is not able to serve our syncing request. " << packet_data.from_node_id_.abridged()
+                     << " peer will be disconnected";
+        peers_state_->set_peer_light_node(peer->getId());
+        disconnect(peer->getId(), dev::p2p::UserReason);
```
Comment (MatusKysel) [inline libraries/core_libs/network/src/tarcap/packets_handlers/status_packet_handler.cpp:63]: do we need to disconnect? I guess it is useless for us, but maybe we could still gossip transactions via RPC back to the network. I do not know what's right

File: libraries/core_libs/storage/include/storage/storage.hpp [hunk 7]
```diff
@@ -148,6 +148,7 @@ class DbStorage : public std::enable_shared_from_this<DbStorage> {
   bool snapshot_enable_ = true;
   uint32_t db_max_snapshots_ = 0;
   std::set<uint64_t> snapshots_;
+  uint64_t light_node_history_ = 0;
```
Comment (MatusKysel) [inline libraries/core_libs/storage/include/storage/storage.hpp:151]: const

File: tests/full_node_test.cpp [hunk 8]
```diff
@@ -955,12 +955,14 @@ TEST_F(FullNodeTest, sync_two_nodes1) {
   auto num_vertices1 = nodes[0]->getDagManager()->getNumVerticesInDag();
   auto num_vertices2 = nodes[1]->getDagManager()->getNumVerticesInDag();
   for (unsigned i = 0; i < SYNC_TIMEOUT; i++) {
-    if (num_vertices1.first > 3 && num_vertices2.first > 3 && num_vertices1 == num_vertices2) break;
+    if (nodes[0]->getTransactionManager()->getTransactionPoolSize() == 0 &&
+        nodes[1]->getTransactionManager()->getTransactionPoolSize() == 0 && num_vertices1 == num_vertices2)
+      break;
```
Comment (MatusKysel) [inline tests/full_node_test.cpp:960]: is this related to your changes?

**Review by MatusKysel — APPROVED** [review]
Maybe you could also create some simple test suite, like we have fullnode -> lightnode? where we could check if everything works consensus wise and deleting old data

File: libraries/core_libs/consensus/src/dag/dag.cpp (hunk 2 above)
Comment (mfrankovi) [reply libraries/core_libs/consensus/src/dag/dag.cpp:423]: I think it is better that it is here under the DagManager mutex. Otherwise there could be a race condition that when we verify block it is not old but when it is inserted to DAG period changed and it could be old.

File: libraries/core_libs/consensus/src/dag/dag.cpp (hunk 3 above)
Comment (mfrankovi) [reply libraries/core_libs/consensus/src/dag/dag.cpp:601]: No, this is not related to light node functionality. This is about expiring old DAG blocks. It applies both for light node and full node

File: libraries/core_libs/network/rpc/Test.cpp (hunk 4 above)
Comment (mfrankovi) [reply libraries/core_libs/network/rpc/Test.cpp:54]: I am not sure how to handle the rpc calls for light node. In current implementation if in a light node if dag block is in the part of period_data that we deleted isDagBlockKnown will return true but getDagBlock will return a null value because the period_data is missing. I am not sure if this is the best way to go.

File: libraries/core_libs/network/src/tarcap/packets_handlers/status_packet_handler.cpp (hunk 5 above)
Comment (mfrankovi) [reply libraries/core_libs/network/src/tarcap/packets_handlers/status_packet_handler.cpp:55]: Fixed

File: libraries/core_libs/network/src/tarcap/packets_handlers/status_packet_handler.cpp (hunk 6 above)
Comment (mfrankovi) [reply libraries/core_libs/network/src/tarcap/packets_handlers/status_packet_handler.cpp:63]: If we are out of sync I think we need to disconnect which will trigger us connecting to other nodes. I want to avoid the state where we are connected only to light nodes

File: libraries/core_libs/storage/include/storage/storage.hpp (hunk 7 above)
Comment (mfrankovi) [reply libraries/core_libs/storage/include/storage/storage.hpp:151]: Fixed

File: tests/full_node_test.cpp (hunk 8 above)
Comment (mfrankovi) [reply tests/full_node_test.cpp:960]: I am not sure what triggered this tests to not pass but I got intermittent failures here where vertices count was 3. I don't understand why we had it to be over count of 3. I replaced this with checking that the transactions pool is empty which means that all the inserted transactions made it to DAG.

File: libraries/config/include/config/config.hpp [hunk 9]
```diff
@@ -71,6 +73,10 @@ struct TestParamsConfig {
 };
 
 struct FullNodeConfig {
+  static const uint32_t kDagExpiryPeriodLimit = 1000;  // Any non finalized dag block with a propose period smaller by
```
Comment (gaoqi8649) [inline libraries/config/include/config/config.hpp:76]: I have a concern on this design. After introduce PBFT blocks with NULL anchor, we may have lots of empty PBFT blocks when there is no transactions generated. The proposal period only increase when finalizing DAG blocks. When generate empty PBFT blocks, proposal period will not increase but finalized period will keep increase. So the proposal period will be much behind the finalized period

File: libraries/cli/include/cli/default_config.hpp (hunk 1 above)
Comment (gaoqi8649) [reply libraries/cli/include/cli/default_config.hpp:20]: I think "light node history" should be a const number and is not changeable. 
Maybe you want add in config for testing to adjust a good number, and remove it later.

File: libraries/config/include/config/config.hpp (hunk 9 above)
Comment (mfrankovi) [reply libraries/config/include/config/config.hpp:76]: Yes, true, I will need to make changes not to count empty pbft blocks.

File: libraries/core_libs/consensus/src/dag/dag.cpp (hunk 2 above)
Comment (MatusKysel) [reply libraries/core_libs/consensus/src/dag/dag.cpp:425]: but it could be in both. This way we could save some CPU if we receive old blocks

File: libraries/config/src/config.cpp [hunk 10]
```diff
@@ -144,6 +146,14 @@ FullNodeConfig::FullNodeConfig(Json::Value const &string_or_object, Json::Value
     network.network_boot_nodes.push_back(node);
   }
 
+  light_node_history = getConfigDataAsUInt(root, {"light_node_history"}, true, 0);
+
+  // Min history for light node is 110% of kDagExpiryPeriodLimit to avoid any race condition around the
+  // kDagExpiryPeriodLimit
+  if (light_node_history > 0 && light_node_history < kDagExpiryPeriodLimit * 1.1) {
+    light_node_history = kDagExpiryPeriodLimit * 1.1;
+  }
```
Comment (gaoqi8649) [inline libraries/config/src/config.cpp:155]: I think it's better to make kDagExpiryPeriodLimit to 1100.
Then:
if (light_node_history != 0) {
  light_node_history = std::min(light_node_history, kDagExpiryPeriodLimit);
}

File: libraries/core_libs/network/src/tarcap/shared_states/peers_state.cpp [hunk 11]
```diff
@@ -134,4 +134,31 @@ bool PeersState::is_peer_malicious(const dev::p2p::NodeID& peer_id) {
   return false;
 }
 
+void PeersState::set_peer_light_node(const dev::p2p::NodeID& peer_id) {
+  light_node_peers_.emplace(peer_id, std::chrono::steady_clock::now());
+}
+
+bool PeersState::is_peer_light_node(const dev::p2p::NodeID& peer_id) {
+  // Peers are marked light nodes for the time defined in conf_.network_peer_light_node_timeout
```
Comment (gaoqi8649) [inline libraries/core_libs/network/src/tarcap/shared_states/peers_state.cpp:142]: I don't understand why define light node peers in timing. 
In this way, some light nodes will be treat as light nodes, and some will be not. 
What is the purpose for that?

File: libraries/core_libs/storage/include/storage/storage.hpp [hunk 12]
```diff
@@ -94,7 +94,7 @@ class DbStorage : public std::enable_shared_from_this<DbStorage> {
     // do not change/move
     COLUMN(default_column);
     // Contains full data for an executed PBFT block including PBFT block, cert votes, dag blocks and transactions
-    COLUMN(period_data);
+    COLUMN_W_COMP(period_data, getIntComparator<uint64_t>());
```
Comment (gaoqi8649) [inline libraries/core_libs/storage/include/storage/storage.hpp:97]: I guess the COLUMN_W_COMP will do saving in order in DB. Am i right?

File: libraries/core_libs/network/src/tarcap/shared_states/peers_state.cpp (hunk 11 above)
Comment (mfrankovi) [reply libraries/core_libs/network/src/tarcap/shared_states/peers_state.cpp:142]: it is possible for a node to change its state and go from light node back to full node

File: libraries/core_libs/storage/include/storage/storage.hpp (hunk 12 above)
Comment (mfrankovi) [reply libraries/core_libs/storage/include/storage/storage.hpp:97]: Yes. This is needed for DeleteRange method in the db to work correctly.

File: libraries/cli/include/cli/default_config.hpp [hunk 13]
```diff
@@ -16,6 +16,8 @@ const char *default_json = R"foo({
   "network_sync_level_size": 25,
```
Comment (JakubFornadel) [inline libraries/cli/include/cli/default_config.hpp:16]: we have now also mainnet_config.hpp, which should be propably adjusted too

File: libraries/config/src/config.cpp [hunk 14]
```diff
@@ -144,6 +146,14 @@ FullNodeConfig::FullNodeConfig(Json::Value const &string_or_object, Json::Value
     network.network_boot_nodes.push_back(node);
   }
 
+  light_node_history = getConfigDataAsUInt(root, {"light_node_history"}, true, 0);
+
+  // Min history for light node is 110% of kDagExpiryPeriodLimit to avoid any race condition around the
+  // kDagExpiryPeriodLimit
+  if (light_node_history > 0 && light_node_history < kDagExpiryPeriodLimit * 1.1) {
```
Comment (JakubFornadel) [inline libraries/config/src/config.cpp:153]: We have a void `FullNodeConfig::validate()` function in this file, which checks values limits, so this check should be probably there so we keep it on the same place + it should throw ConfigException rather than overriding the value so the node does not even start up and enforce user to setup this value properly in case he put there invalid number

File: libraries/core_libs/consensus/src/dag/dag.cpp [hunk 15]
```diff
@@ -584,25 +596,89 @@ uint DagManager::setDagBlockOrder(blk_hash_t const &new_anchor, uint64_t period,
...
+  updateFrontier();
 
   LOG(log_nf_) << "Set new period " << period << " with anchor " << new_anchor;
 
   return dag_order_set.size();
 }
 
+void DagManager::handleExpiredDagBlocksTransactions(std::vector<trx_hash_t> &transactions_from_expired_dag_blocks) {
+  std::unordered_set<trx_hash_t> transactions_from_expired_dag_blocks_to_remove;
+  // Only transactions that need to be moved to memory pool are the one that are not yet finalized and are not part of
```
Comment (JakubFornadel) [inline libraries/core_libs/consensus/src/dag/dag.cpp:631]: are the ones

File: libraries/core_libs/consensus/src/dag/dag.cpp [hunk 16]
```diff
@@ -584,25 +596,89 @@ uint DagManager::setDagBlockOrder(blk_hash_t const &new_anchor, uint64_t period,
...
   anchor_ = new_anchor;
   period_ = period;
+  updateFrontier();
 
   LOG(log_nf_) << "Set new period " << period << " with anchor " << new_anchor;
 
   return dag_order_set.size();
 }
 
+void DagManager::handleExpiredDagBlocksTransactions(std::vector<trx_hash_t> &transactions_from_expired_dag_blocks) {
```
Comment (JakubFornadel) [inline libraries/core_libs/consensus/src/dag/dag.cpp:629]: missing two const ? 
`void DagManager::handleExpiredDagBlocksTransactions(const std::vector<trx_hash_t> &transactions_from_expired_dag_blocks)  const `

File: libraries/core_libs/consensus/src/dag/dag.cpp [hunk 17]
```diff
@@ -584,25 +596,89 @@ uint DagManager::setDagBlockOrder(blk_hash_t const &new_anchor, uint64_t period,
   assert(dag_order_set.count(new_anchor));
   addToDag(new_anchor, blk_hash_t(), vec_blk_t(), 0, true);
 
+  std::pair<std::unordered_map<blk_hash_t, std::shared_ptr<DagBlock>>, std::vector<trx_hash_t>>
+      expired_dag_blocks_to_remove;
```
Comment (JakubFornadel) [inline libraries/core_libs/consensus/src/dag/dag.cpp:600]: This whole std::pair seems overcomplicated and could be simplified, which would also improved readability:

Instead of 1 pair, you could declre two separate containers with proper names:
```
std::unordered_map<blk_hash_t, std::shared_ptr<DagBlock>> ct1;
std::vector<trx_hash_t> ct2;
```

Read comments below, which are related

File: libraries/core_libs/consensus/src/dag/dag.cpp [hunk 18]
```diff
@@ -584,25 +596,89 @@ uint DagManager::setDagBlockOrder(blk_hash_t const &new_anchor, uint64_t period,
...
-        auto dag_block = dag_blk_mgr_->getDagBlock(blk);
+    for (auto &blk_hash : v.second) {
+      if (dag_order_set.count(blk_hash) == 0) {
+        auto dag_block = dag_blk_mgr_->getDagBlock(blk_hash);
         auto pivot_hash = dag_block->getPivot();
-        addToDag(blk, pivot_hash, dag_block->getTips(), dag_block->getLevel(), false);
+
+        if (validateBlockNotExpired(dag_block, expired_dag_blocks_to_remove)) {
+          addToDag(blk_hash, pivot_hash, dag_block->getTips(), dag_block->getLevel(), false);
+        }
```
Comment (JakubFornadel) [inline libraries/core_libs/consensus/src/dag/dag.cpp:609]: this could be refactored to:

if (validateBlockNotExpired(dag_block, ct1)) {
  addToDag(blk_hash, pivot_hash, dag_block->getTips(), dag_block->getLevel(), false);
} else {
 //  add dag_block->getTrxs() to ct2
}

File: libraries/core_libs/consensus/src/transaction_manager/transaction_manager.cpp [hunk 19]
```diff
@@ -278,6 +278,20 @@ void TransactionManager::updateFinalizedTransactionsStatus(SyncBlock const &sync
   db_->saveStatusField(StatusDbField::TrxCount, trx_count_);
 }
 
+void TransactionManager::moveNonFinalizedTransactionsToTransactionsPool(std::unordered_set<trx_hash_t> &&transactions) {
+  // !!! There is no lock because it is called under std::unique_lock trx_lock(trx_mgr_->getTransactionsMutex());
+  auto write_batch = db_->createWriteBatch();
+  for (auto const &trx_hash : transactions) {
+    db_->removeTransactionToBatch(trx_hash, write_batch);
```
Comment (JakubFornadel) [inline libraries/core_libs/consensus/src/transaction_manager/transaction_manager.cpp:285]: should be this inside `if (trx != nonfinalized_transactions_in_dag_.end()) {` ?

File: libraries/core_libs/network/src/tarcap/shared_states/peers_state.cpp (hunk 11 above)
Comment (JakubFornadel) [reply libraries/core_libs/network/src/tarcap/shared_states/peers_state.cpp:142]: to go from light node to full node, user have to change config and restart the node, in which case the node will create new connections as full node, so `network_peer_light_node_timeout` seems like we could delete it ?

File: libraries/cli/include/cli/default_config.hpp [hunk 20]
```diff
@@ -16,6 +16,8 @@ const char *default_json = R"foo({
...
   "network_packets_processing_threads": 14,
   "network_peer_blacklist_timeout" : 600,
+  "network_peer_light_node_timeout" : 600,
+  "light_node_history" : 0,
```
Comment (JakubFornadel) [inline libraries/cli/include/cli/default_config.hpp:20]: This is one of the very few config parameters that users might actually touch. I would suggest to make it more user friendly and have here flag `"is_light_node" : true/false` + `"light_node_history"`, which would be set to some default value and users would need only change is_light_node param

File: libraries/core_libs/network/src/tarcap/packets_handlers/status_packet_handler.cpp [hunk 21]
```diff
@@ -51,6 +52,17 @@ void StatusPacketHandler::process(const PacketData& packet_data, const std::shar
     auto const node_major_version = (*it++).toInt<unsigned>();
     auto const node_minor_version = (*it++).toInt<unsigned>();
     auto const node_patch_version = (*it++).toInt<unsigned>();
+    auto const node_history = (*it++).toInt<uint64_t>();
+
+    // If this is a light node and it cannot serve our sync request disconnect from it
+    if (node_history > 0) {
+      if (pbft_synced_period + node_history < peer_pbft_chain_size) {
```
Comment (JakubFornadel) [inline libraries/core_libs/network/src/tarcap/packets_handlers/status_packet_handler.cpp:59]: I dont think that this s enough. What if I am light node and when I connected to you this condition did not pass so everything is ok but for some reason, my node get out of sync(for any reason - connection problems, etc...). You would try to get sync data from me but I dont have them. I think we have to handle this + we should probably also save  node_history for each light_node peer so we actually know if and how much data can we ask from him

File: libraries/core_libs/network/src/tarcap/packets_handlers/status_packet_handler.cpp [hunk 22]
```diff
@@ -51,6 +52,17 @@ void StatusPacketHandler::process(const PacketData& packet_data, const std::shar
     auto const node_major_version = (*it++).toInt<unsigned>();
     auto const node_minor_version = (*it++).toInt<unsigned>();
     auto const node_patch_version = (*it++).toInt<unsigned>();
+    auto const node_history = (*it++).toInt<uint64_t>();
```
Comment (JakubFornadel) [inline libraries/core_libs/network/src/tarcap/packets_handlers/status_packet_handler.cpp:55]: if we are going to use "is_light_node" flag in config, I would use it also here + light_node_history

File: libraries/core_libs/consensus/src/dag/dag.cpp (hunk 2 above)
Comment (mfrankovi) [reply libraries/core_libs/consensus/src/dag/dag.cpp:425]: I have put it now in both

File: libraries/config/include/config/config.hpp (hunk 9 above)
Comment (mfrankovi) [reply libraries/config/include/config/config.hpp:76]: Implemented

File: libraries/config/src/config.cpp (hunk 10 above)
Comment (mfrankovi) [reply libraries/config/src/config.cpp:155]: this part of code changed, please review again

File: libraries/cli/include/cli/default_config.hpp (hunk 13 above)
Comment (mfrankovi) [reply libraries/cli/include/cli/default_config.hpp:16]: Fixed

File: libraries/core_libs/consensus/src/dag/dag.cpp (hunk 16 above)
Comment (mfrankovi) [reply libraries/core_libs/consensus/src/dag/dag.cpp:629]: Fixed

File: libraries/core_libs/consensus/src/dag/dag.cpp (hunk 17 above)
Comment (mfrankovi) [reply libraries/core_libs/consensus/src/dag/dag.cpp:600]: Refactored

File: libraries/core_libs/consensus/src/transaction_manager/transaction_manager.cpp (hunk 19 above)
Comment (mfrankovi) [reply libraries/core_libs/consensus/src/transaction_manager/transaction_manager.cpp:285]: Yes, good catch. Fixed

File: libraries/cli/include/cli/default_config.hpp (hunk 20 above)
Comment (mfrankovi) [reply libraries/cli/include/cli/default_config.hpp:20]: Done

File: libraries/core_libs/network/src/tarcap/packets_handlers/status_packet_handler.cpp (hunk 21 above)
Comment (mfrankovi) [reply libraries/core_libs/network/src/tarcap/packets_handlers/status_packet_handler.cpp:59]: I believe that any scenario like this is already handled and we would disconnect from any node that is not able to serve us data we need. Maybe we need additional ways to detect that we are not syncing and try to disconnect from other nodes but I do not think this is related to light node functionality.

File: libraries/config/src/config.cpp [hunk 23]
```diff
@@ -144,6 +146,11 @@ FullNodeConfig::FullNodeConfig(Json::Value const &string_or_object, Json::Value
     network.network_boot_nodes.push_back(node);
   }
 
+  is_light_node = getConfigDataAsBoolean(root, {"is_light_node"}, true, false);
+  if (is_light_node) {
+    light_node_history = getConfigDataAsUInt(root, {"light_node_history"}, true, kDefaultLightNodeHistory);
```
Comment (gaoqi8649) [inline libraries/config/src/config.cpp:151]: You removed light_node_history in config file

File: libraries/core_libs/network/src/tarcap/shared_states/peers_state.cpp (hunk 11 above)
Comment (JakubFornadel) [reply libraries/core_libs/network/src/tarcap/shared_states/peers_state.cpp:142]: exactly, this `light_node` flag should be inside Peer class as once the new node connects it is either light or full node. It cannot change it's state without restarting the node. So let's put it into the Peer class and then you dont need no timeout either

File: libraries/core_libs/network/src/tarcap/packets_handlers/status_packet_handler.cpp (hunk 21 above)
Comment (JakubFornadel) [reply libraries/core_libs/network/src/tarcap/packets_handlers/status_packet_handler.cpp:66]: hmm I think the solution is pretty easy to implement. Check my other comment first. I would do this:

- add light_node flag to Peer class
- add optional<light_node_history> to Peer class
- any time we ask for sync data, we check if the peer is light node(we will node as it will be in Peer class). If it is light node we will check also num of pbft blocks he keeps and compare it to what we are asking for. If he does not have such long history, we will skip for another peer

This check will stay here + the functionality from above will make it robust so in any scenario (which we might not even know about) we will not ask for sync data from a light node, which he wont ba able to serve

File: libraries/core_libs/consensus/src/pbft/pbft_chain.cpp [hunk 24]
```diff
@@ -107,6 +114,9 @@ void PbftChain::updatePbftChain(blk_hash_t const& pbft_block_hash, bool null_anc
   size_++;
   if (!null_anchor) {
     non_empty_size_++;
+    if ((dag_expiry_period_ == 0 && non_empty_size_ > kDagExpiryPeriodLimit) || dag_expiry_period_ > 0) {
```
Comment (gaoqi8649) [inline libraries/core_libs/consensus/src/pbft/pbft_chain.cpp:117]: This should not right. Like non_empty_size_ is 2000 and faucet down at the moment. There may still some DAG blocks in the last DAG level as sibling for non finalized. If no new DAG block generate, these DAG blocks will be treat as expired.

File: libraries/config/src/config.cpp (hunk 23 above)
Comment (JakubFornadel) [reply libraries/config/src/config.cpp:151]: yeah I think it was because of my comment with flag. So the idea with flag is that most people, who will run light node wont really configure light_node_history so it is much more user friendly to have the light_node flag, which they can set to true/false and use default value for light_node_history.

But we still can and probably should have also this parameter for setting the light_node_history -> Some advanced users might run some kind of application where they need full data for the past N days for example so they might set this history to some bigger number as the default value we use...

Comment (mfrankovi) [reply libraries/config/src/config.cpp:151]: this is added as an optional parameter which is not present in the default config files but it is supported, I think this kind of more advanced params do not need to be in the config file but they can be added manually if advanced user wants to add them.

File: libraries/core_libs/network/src/tarcap/packets_handlers/status_packet_handler.cpp (hunk 21 above)
Comment (mfrankovi) [reply libraries/core_libs/network/src/tarcap/packets_handlers/status_packet_handler.cpp:66]: I can try to implement this but I still believe that we should disconnect from light nodes if we are out of sync and light nodes cannot serve our sync requests. There is always a chance that most/all of the connected nodes are light nodes and I do not see any point in being connected with light node if we are out of sync. If we do not have some timeout/state outside of the Peer class than we might get into some state where we keep connecting over and over again to the same peer so I do not see how this can work with keeping this only in Peer class

**Review by gaoqi8649 — APPROVED** [review]

File: libraries/core_libs/network/src/tarcap/packets_handlers/status_packet_handler.cpp (hunk 21 above)
Comment (JakubFornadel) [reply libraries/core_libs/network/src/tarcap/packets_handlers/status_packet_handler.cpp:66]: true, to solve this without any complicated timers: once we get into the `restartPbftSyncing` function we can go through the connected peers and if we find out that none of the connected peers are able to serve us the data(because they are light nodes), we would simply disconnect all of them and we would not be able to create a new  connection with them due the condition in initial status packet

File: libraries/cli/include/cli/default_config.hpp [hunk 25]
```diff
@@ -16,6 +16,8 @@ const char *default_json = R"foo({
   "network_sync_level_size": 25,
   "network_packets_processing_threads": 14,
   "network_peer_blacklist_timeout" : 600,
+  "network_peer_light_node_timeout" : 600,
+  "is_light_node" : false,
```
Comment (JakubFornadel) [inline libraries/cli/include/cli/default_config.hpp:20]: are we going to use also `light_node_history` config parameter together with `is_light_node` flag ?

- is_light_node is just easy way to say for most of the users if they wanna run light/full node through true/false flag
- light_node_history would be set to some default value that most of the users dont need to touch at all, but if someone builds an app on top of our blockchain and needs to run local node with lets say 1 week of history, he would have a possibility to set custom history for his light node

File: libraries/core_libs/network/src/tarcap/shared_states/peers_state.cpp (hunk 11 above)
Comment (mfrankovi) [reply libraries/core_libs/network/src/tarcap/shared_states/peers_state.cpp]: Implemented

File: libraries/core_libs/network/src/tarcap/packets_handlers/get_pbft_sync_packet_handler.cpp [hunk 26]
```diff
@@ -39,6 +42,13 @@ void GetPbftSyncPacketHandler::process(const PacketData &packet_data,
     throw MaliciousPeerException(err_msg.str());
   }
 
+  if (is_light_node_ && height_to_sync + light_node_history_ <= my_chain_size) {
+    std::ostringstream err_msg;
+    err_msg << "Peer " << packet_data.from_node_id_ << " request syncing period start at " << height_to_sync
+            << ". Light node does not have the data " << my_chain_size;
+    throw MaliciousPeerException(err_msg.str());
```
Comment (gaoqi8649) [inline libraries/core_libs/network/src/tarcap/packets_handlers/get_pbft_sync_packet_handler.cpp:49]: I think sender should not know the peer is a light node and don't have the block before sending request. Do we need to mark sender as malicous here?

Comment (mfrankovi) [reply libraries/core_libs/network/src/tarcap/packets_handlers/get_pbft_sync_packet_handler.cpp:49]: Sender should know this since it is being sent as part of status message
